formula: add `to_recursive_bottle_hash` method

Rylan12 opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

This PR adds the `Formula#to_bottle_hash` method to be used to create bottle JSON files for #11426.

This will be used in Homebrew/formulae.brew.sh to create the `/api/bottles` API.

Click these links for gists showing the output of this method (when converted to JSON) is for these formulae: [`hello`](https://gist.github.com/Rylan12/bf0adea0d317697a8f16dfc12637ab1c), [`youtube-dl`](https://gist.github.com/Rylan12/5756038a327159cf7f449f44f7406de2)

Rylan12 commented:
I'm going to cross-post what I commented in https://github.com/Homebrew/brew/pull/11426#issuecomment-854018365 because it's more relevant here:

Another question, how to handle dependencies that are only needed on some platforms or in some situations (e.g. [`asciidoctor`](https://github.com/Homebrew/homebrew-core/blob/7033ad8c4acca1e821274c0f3323b0a4b7e8971e/Formula/asciidoctor.rb#L15), [`sparse`](https://github.com/Homebrew/homebrew-core/blob/7033ad8c4acca1e821274c0f3323b0a4b7e8971e/Formula/sparse.rb#L16), or [`cocoapods`](https://github.com/Homebrew/homebrew-core/blob/7033ad8c4acca1e821274c0f3323b0a4b7e8971e/Formula/cocoapods.rb#L17))?

The biggest problem, I think, is figuring out how to detect them automatically to export a `Formula` into a JSON file. Right now, this would return a different result depending on the platform on which it's being run. How do we detect which dependencies are needed on platforms that aren't being used currently?

One possibility is to change the DSL to handle the conditions rather than using an `if`. I'm not sure that's the best way, though, as it needs to change countless formulae.

Another option might be to create a new brew command to generate these JSON files that automatically checks for dependencies on all systems by temporarily overriding the value of `MacOS::version`, `Hardware::CPU::arm?`, etc.

---

Separately, how should dependencies be stored in the JSON file? I'm now thinking that since most of the time they'll be the same for all bottle tags, they should only be listed once in the top-level of the JSON file and can just specify which versions they're needed for. For example:

```json
{
  "bottles": {
    "big_sur": { ... },
    "catalina": { ... }
  },
  "dependencies": [
    {
      "big_sur": {
        ...,
        "clang_build_version": ">= 1000"
      },
      "catalina": {
        ...,
        "arm": true
      },
    },
    {
      "big_sur": { ... }
    }
  ]
}
```

In the above example, `...` is where the `url` and `sha256` would go. You can see that on Big Sur, both dependencies are needed as they both have a `big_sur` bottle tag. We could also include this as an option like I did for `arm` and `clang_build_version` which are both items of the bottle objects that could be interpreted as conditions to be checked when choosing which dependencies.

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1908,6 +1908,27 @@ def to_hash
 
+  # @api private
+  # Generate a hash to be used to install a formula from a JSON file
+  def to_bottle_hash(top_level: true)
```
might want to rename given mild confusion with `bottle_hash` method below.

Rylan12 replied:
Any ideas?

Rylan12 replied:
I went with `to_recursive_bottle_hash` for now

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1908,6 +1908,27 @@ def to_hash
+    bottles = bottle["files"].map do |tag, file|
+      info = {
+        "url"    => file["url"],
+        "sha256" => file["sha256"],
```
I think it's worth omitting these for homebrew/core and linuxbrew/core,

MikeMcQuaid reviewed (commented):
Looks good so far! Would like to have this have a way of actually being used before it's merged e.g. `brew info --json --bottle` or something.

Rylan12 commented:
@MikeMcQuaid can you comment on the [above issue](https://github.com/Homebrew/brew/pull/11488#issuecomment-854033374) with dependencies that are hidden behind conditionals?

Rylan12 commented:
I've added a `--bottle` flag to `brew info --json` (both `v1` and `v2`) that outputs this JSON. I've also removed the `sha256` for core formulae. Note that if you run this on a non-core formula that depends on a core formula, the core formula bottles that are listed in the `dependencies` section will also not include the `sha256`.

MikeMcQuaid commented on Library/Homebrew/formula.rb:
```diff
@@ -1908,6 +1908,30 @@ def to_hash
+
+    bottles = bottle["files"].map do |tag, file|
+      info = { "url" => file["url"] }
+      info["sha256"] = file["sha256"] unless tap.name == "homebrew/core"
```
```suggestion
      info["sha256"] = file["sha256"] if tap.name != "homebrew/core"
```

MikeMcQuaid reviewed (approved):
Tiny tweak, otherwise 👍🏻 👏🏻

MikeMcQuaid commented:
> Note that if you run this on a non-core formula that depends on a core formula, the core formula bottles that are listed in the `dependencies` section will also not include the `sha256`.

This seems ideal, thanks 👍🏻

MikeMcQuaid commented:
> I've added a `--bottle` flag to `brew info --json` (both `v1` and `v2`) that outputs this JSON.

I'd be tempted to only add this to `v2` (but don't feel strongly).

Rylan12 commented:
> > I've added a `--bottle` flag to `brew info --json` (both `v1` and `v2`) that outputs this JSON.
> 
> I'd be tempted to only add this to `v2` (but don't feel strongly).

I don't feel strongly about all about this. I don't really know much about why we have `v1` and `v2` although I'd guess it's just a backwards compatibility thing. The benefit to having it for `v1` as well is that using just `--json` instead of `--json=v2` will work. Since the `--bottle` flag is being added here, I don't think this causes a compatibility issue that we need to worry about.

Rylan12 commented:
@MikeMcQuaid I'm going to merge so I can push forward a bit on the other stuff that this is blocking (and since you said you don't feel strongly). That part can always be reverted.

MikeMcQuaid commented:
@Rylan12 works for me 👍🏻
